### ISSUE TITLE
Running binary in multiple binary-project does not work

### ISSUE THREAD

**Issue by skade** [rust-analyzer/rust-analyzer#1940]
When clicking the `Run` button on top of a main function, rust-analyzer will trigger `cargo run`.

This does not work in multi-binary projects where the binary must be specified.

For example, in a project with the following layout:

```
|
|- bin
  |- client.rs
  |- server.rs
```

The output is:

```sh
> Executing task: cargo run <

error: `cargo run` could not determine which binary to run. Use the `--bin` option to specify a binary, or the `default-run` manifest key.
available binaries: client, server
The terminal process terminated with exit code: 101

Terminal will be reused by tasks, press any key to close it.
```

**Comment by kjeremy** [comment]
The entry point for this is on the LSP is in handle_code_lens.rs here: https://github.com/rust-analyzer/rust-analyzer/blob/ebfba461123928363da55e75195d76314c51fa7e/crates/ra_lsp_server/src/main_loop/handlers.rs#L636

Above that `main` is detected here: https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ra_ide_api/src/runnables.rs#L44

The typsecript portion is https://github.com/rust-analyzer/rust-analyzer/blob/526a6aba104a32eb9f0f5a65232783d5570c35d5/editors/code/src/commands/runnables.ts

We probably need to determine the binary on the LSP side and thread it through to the typescript.

**Comment by matklad** [comment]
My hypothesis is that the problem is elsewhere: I think that either server or client were added after the rust-analyzer was started, so we didn't get them in `cargo metadata` from the start. That is, the bug is that we don't support dynamic reconfiguration (and don't even issue a notice that). 

And I think it is maybe time to start working on dynamic reconfiguration: this will also transition vfs to the on-deman file loading

**Comment by marcogroppo** [comment]
I reproduced the bug with a new lib crate and I can confirm that it disappeared after closing and re-opening VSCode. The first time you create a `main` function in a lib crate `Run` works "by accident" because `cargo run --bin` does not need any additional parameters... but RA isn't returning the `CargoTargetSpec`.

**Comment by matklad** [comment]
Closing in favor of https://github.com/rust-analyzer/rust-analyzer/issues/1968
